collection+xattrs.go

```diff
@@ -618,16 +619,74 @@ func removeUserXattrs(xattrs semiParsedXattrs) {
 	}
 }
 
+// macroExpand takes the macro and assigned 'value' the correct value based on what the macro is provided (this is gocb macro format)
+func (e *event) macroExpand(value string, casServerFormat string, bucketUUID string) string {
+	if value == "\"${Mutation.CAS}\"" {
+		value = casServerFormat
+	} else if value == "\"${Mutation.value_crc32c}\"" {
+		value = encodedCRC32c(e.value)
+	} else {
+		// if the value isn't to be macro expanded to CAS or CRC hash then it is a source ID
+		value = bucketUUID
+	}
+	return value
+}
+
 // Sets JSON properties "cas" to the given `cas`, and "value_crc" to CRC checksum of `docValue`.
-func (e *event) expandSyncXattrMacros(xattr any) {
+func (e *event) expandSyncXattrMacros(xattr any, bucketUUID string, mutateSpec *sgbucket.MutateInOptions) error {
+	var err error
+	if mutateSpec == nil {
+		emptySpec := sgbucket.MutateInOptions{}
+		mutateSpec = &emptySpec
+	}
+
 	if xattrMap, ok := xattr.(map[string]any); ok {
-		// For some reason Server encodes `cas` as 8 hex bytes in little-endian order...
+		// create server format of cas value
 		casBytes := make([]byte, 8)
 		binary.LittleEndian.PutUint64(casBytes, e.cas)
-		xattrMap["cas"] = fmt.Sprintf("0x%x", casBytes)
+		casServerFormat := fmt.Sprintf("0x%x", casBytes)
 
-		xattrMap["value_crc32c"] = encodedCRC32c(e.value)
+		// loop through provided mutate in spec and macro expand and insert into xattr
+		for _, v := range mutateSpec.Spec {
+			str := fmt.Sprint(v.Value)
+			v.Value = e.macroExpand(str, casServerFormat, bucketUUID)
```

Comment:
`macroExpand` is only called here, and I think it would be clearer if its source were just inlined into the call-site.